src/node/configuration_tracker.h

```diff
@@ -0,0 +1,413 @@
+          ss << id;
+          auto nit = nodes.find(id);
+          if (nit == nodes.end() && id != node_id)
+            ss << "!";
+          else if (nit != nodes.end() && nit->second.catching_up)
+            ss << "*";
+        }
+        ss << "]@" << c.idx << " ";
+      }
+      return ss.str();
+    }
+
+    bool record_promotion(const NodeId& node_id)
+    {
+      LOG_INFO_FMT(
+        "Promoting {} to TRUSTED as it has caught up with us", node_id);
+
+      // Serialize request object
+      PromoteNodeToTrusted::In ps = {node_id};
+
+      http::Request request(fmt::format(
+        "/{}/{}", ccf::get_actor_prefix(ccf::ActorsType::nodes), "promote"));
+      request.set_header(
+        http::headers::CONTENT_TYPE, http::headervalues::contenttype::JSON);
+
+      auto body = serdes::pack(ps, serdes::Pack::Text);
+      request.set_body(&body);
+
+      auto node_cert_der = crypto::cert_pem_to_der(node_cert);
+      const auto key_id = crypto::Sha256Hash(node_cert_der).hex_str();
+
+      http::sign_request(request, node_sign_kp, key_id);
+
+      std::vector<uint8_t> packed = request.build_request();
+
+      // Submit request
+      auto node_session = std::make_shared<enclave::SessionContext>(
+        enclave::InvalidSessionId, node_cert.raw());
+      auto ctx = enclave::make_rpc_context(node_session, packed);
+
+      const auto actor_opt = http::extract_actor(*ctx);
+      if (!actor_opt.has_value())
+      {
+        throw std::logic_error("Unable to get actor");
+      }
+
+      const auto actor = rpc_map->resolve(actor_opt.value());
+      auto frontend_opt = this->rpc_map->find(actor);
+      if (!frontend_opt.has_value())
+      {
+        throw std::logic_error(
+          "RpcMap::find returned invalid (empty) frontend");
+      }
+
+      auto frontend = frontend_opt.value();
+      frontend->process(ctx);
+
+      auto rs = ctx->get_response_status();
+
+      if (rs != HTTP_STATUS_OK)
+      {
+        auto ser_res = ctx->serialise_response();
+        std::string str((char*)ser_res.data(), ser_res.size());
+        LOG_FAIL_FMT("Promotion request failed: {}", str);
+      }
+
+      return rs == HTTP_STATUS_OK;
+    }
+
+    void check_for_promotions(const kv::Configuration::Nodes& config)
+    {
+      for (const auto& [id, info] : nodes)
+      {
+        LOG_DEBUG_FMT("Promo check node {}: {}", id, info.catching_up);
+      }
+
+      for (const auto& [id, info] : config)
+      {
+        LOG_DEBUG_FMT("In config: {}: {}", id, info.catching_up);
+        if (!info.catching_up)
+        {
+          auto nit = nodes.find(id);
+          if (nit != nodes.end() && nit->second.catching_up)
+          {
+            LOG_DEBUG_FMT("Observing promotion: {}", id);
+            nit->second.catching_up = false;
+          }
+        }
+      }
+    }
+
+    static void promote_cb(
+      ConfigurationTracker& configuration_tracker,
+      NodeId from,
+      TxID txid,
+      TxID primary_txid)
+    {
```

Comment:
this is already occurring in an async callback why is another needed?

Reply:
Another what? Am I checking some condition twice or something like that?